### PR TITLE
[AIRFLOW-6530] Allow Custom Statsd Client

### PR DESCRIPTION
**Pull request by envious**
This change allows Airflow users to utilise their own custom Statsd client as discussed [here](https://issues.apache.org/jira/browse/AIRFLOW-6530).

Many companies have their own custom Statsd clients and this change should allow for easier adoption of Airflow by large corporations.

Example usage based off of this change:

1) User adds a module path to their airflow.cfg file next to the 'statsd_custom_client_path' key in the scheduler section of the config.

As such:
`statsd_custom_client_path = company.statsdclient.customclient`

2) Ensure the newly added client exists on the PYTHONPATH.

---
Issue link: [AIRFLOW-6530](https://issues.apache.org/jira/browse/AIRFLOW-6530)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/config_templates/config.yml [hunk 1]
```diff
@@ -1314,6 +1314,15 @@
...
       default: ""
+    - name: statsd_custom_client_path
+      description: |
+        If you want to utilise your own custom Statsd client set the relevant
+        module path below. Note: The module path must exist on your PYTHONPATH
+        for Airflow to pick it up
+      version_added: ~
+      type: string
+      example: ~
+      default: ""
```
Comment (kaxil) [inline airflow/config_templates/config.yml]:
```suggestion
      default: ~
```

As it is optional option so let's keep it as **null**

File: airflow/config_templates/config.yml [hunk 2]
```diff
@@ -1314,6 +1314,15 @@
       type: string
       example: ~
       default: ""
+    - name: statsd_custom_client_path
+      description: |
+        If you want to utilise your own custom Statsd client set the relevant
+        module path below. Note: The module path must exist on your PYTHONPATH
```
Comment (kaxil) [inline airflow/config_templates/config.yml]:
```suggestion
        module path below. 
        Note: The module path must exist on your PYTHONPATH for Airflow to pick it up
```

File: airflow/config_templates/default_airflow.cfg [hunk 3]
```diff
@@ -631,6 +631,11 @@ statsd_prefix = airflow
 # start with the elements of the list (e.g: scheduler,executor,dagrun)
 statsd_allow_list =
 
+# If you want to utilise your own custom Statsd client set the relevant module
+# path below
+# Note: The module path must exist on your PYTHONPATH for Airflow to pick it up
+statsd_custom_client_path =
```
Comment (kaxil) [inline airflow/config_templates/default_airflow.cfg]:
```suggestion
# statsd_custom_client_path =
```

**Comment by envious** [comment]
@kaxil thanks for your suggested changes, i've applied them. I'll figure out how to best test these changes today. Any clues?

**Comment by kaxil** [comment]
@envious - You can add another test at https://github.com/apache/airflow/blob/master/tests/test_stats.py by creating a new Statsd client and passing it using `conf_vars` like 

https://github.com/apache/airflow/blob/d7d2794d05f44b6dd7edf2d44633cba9951e3db0/tests/utils/test_email.py#L90-L98

and asserting that the client is the same that you were expecting.

**Comment by mik-laj** [comment]
> Many companies have their own custom Statsd clients and this change should allow for easier adoption of Airflow by large corporations.

Can you say something more about that? Do you mean any public software or services?

**Comment by craigrosie** [comment]
While tests are being worked on for this, do you have any advice on how/if we can get this "backported" to the next 1.10 release?

**Comment by kaxil** [comment]
> While tests are being worked on for this, do you have any advice on how/if we can get this "backported" to the next 1.10 release?

marked it for 1.10.10. We will cherry-pick and backport your changes after merging it to master

**Comment by craigrosie** [comment]
Hey @kaxil / @mik-laj, are you able to give this another review now that we've added tests? 😄 Thank you in advance!

File: airflow/config_templates/config.yml [hunk 4]
```diff
@@ -1401,6 +1401,16 @@
       type: string
       example: ~
       default: ""
+    - name: statsd_custom_client_path
+      description: |
+        If you want to utilise your own custom Statsd client set the relevant
+        module path below.
+        Note: The module path must exist on your PYTHONPATH for Airflow to pick it up
+        for Airflow to pick it up
```
Comment (kaxil) [inline airflow/config_templates/config.yml]: Duplicate wording

File: airflow/stats.py [hunk 5]
```diff
@@ -200,10 +200,35 @@ def __init__(self, *args, **kwargs):
                     self.__class__.instance = DummyStatsLogger()
             except (socket.gaierror, ImportError) as e:
                 log.warning("Could not configure StatsClient: %s, using DummyStatsLogger instead.", e)
+                self.__class__.instance = DummyStatsLogger()
```
Comment (kaxil) [inline airflow/stats.py:203]: Why this change?

Comment (craigrosie) [reply airflow/stats.py:203]: I think I found a bug when adding tests for the new functionality. The log message above the line I added states that the class used for the statsd client should fall back to `DummyStatsLogger` if the defined statsd client class cannot be imported, or cannot connect to a statsd instance, but this never happens, and `self.__class__.instance` remains set to the default value of `None`

For example, if I try to load an invalid custom statsd class using the functionality that @envious added, triggering an `ImportError`, when I then try and call `Stats.incr()` (which should still work because the underlying class has fallen back to `DummyStatsLogger`), I get errors like:
```python
cls = <class 'airflow.stats.Stats'>, name = 'incr'

    def __getattr__(cls, name):
>       return getattr(cls.instance, name)
E       AttributeError: 'NoneType' object has no attribute 'incr'

airflow/stats.py:188: AttributeError
```

File: airflow/config_templates/config.yml (hunk 4 above)
Comment (craigrosie) [reply airflow/config_templates/config.yml]: Thanks, fixed! 😄

File: airflow/config_templates/default_airflow.cfg [hunk 6]
```diff
@@ -674,6 +674,12 @@ statsd_datadog_enabled = False
 # List of datadog tags attached to all metrics(e.g: key1:value1,key2:value2)
 statsd_datadog_tags =
 
+# If you want to utilise your own custom Statsd client set the relevant
+# module path below.
+# Note: The module path must exist on your PYTHONPATH for Airflow to pick it up
+# for Airflow to pick it up
```
Comment (kaxil) [inline airflow/config_templates/default_airflow.cfg]:
```suggestion
```

File: airflow/stats.py [hunk 7]
```diff
@@ -200,10 +200,35 @@ def __init__(self, *args, **kwargs):
...
 
     def get_statsd_logger(self):
-        from statsd import StatsClient
-        statsd = StatsClient(
+        if conf.getboolean('scheduler', 'statsd_on'):
+
```
Comment (kaxil) [inline airflow/stats.py]:
```suggestion
```

File: airflow/stats.py [hunk 8]
```diff
@@ -200,10 +200,35 @@ def __init__(self, *args, **kwargs):
...
+            from statsd import StatsClient
+
+            if conf.has_option('scheduler', 'statsd_custom_client_path'):
+
```
Comment (kaxil) [inline airflow/stats.py]:
```suggestion
```

File: airflow/stats.py [hunk 9]
```diff
@@ -200,10 +200,35 @@ def __init__(self, *args, **kwargs):
...
+                custom_statsd_module_path = conf.get('scheduler', 'statsd_custom_client_path')
+
+                try:
+                    stats_class = import_string(custom_statsd_module_path)
+
```
Comment (kaxil) [inline airflow/stats.py]:
```suggestion
```

**Review by kaxil — APPROVED** [review]
